### PR TITLE
Adding response headers in the RLS server 

### PR DESCRIPTION
**Pull request by chirino**
so that it implements https://datatracker.ietf.org/doc/id/draft-polli-ratelimit-headers-03.html

See issue #156

### CODE REVIEW

**Comment by chirino** [comment]
@alexsnaps could ya review this one?

File: limitador/src/storage/in_memory.rs [hunk 1]
```diff
@@ -31,22 +31,40 @@ impl CounterStorage for InMemoryStorage {
...
-            if !Self::counter_is_within_limits(counter, stored_counters.get(counter), delta) {
-                return Ok(Authorization::Limited(
-                    counter.limit().name().map(|n| n.to_owned()),
-                ));
+        if load_counters {
+            let mut first_limited = None;
+            for counter in counters.iter_mut() {
+                let remaining =
+                    *stored_counters.get(counter).unwrap_or(&counter.max_value()) - delta;
+                counter.set_remaining(remaining);
```
Comment (alexsnaps) [inline limitador/src/storage/in_memory.rs:46]: We're not updating the `expires_in` duration on this store... Not that I know how we'd do that. `TtlCache` doesn't seem to give us access to `InternalEntry.expiration` 🤔 
But I think we should get rid of it anyways... see #71

File: limitador/src/storage/mod.rs [hunk 2]
```diff
@@ -121,10 +121,12 @@ impl Storage {
 
     pub fn check_and_update(
         &self,
-        counters: HashSet<Counter>,
+        counters: &mut Vec<Counter>,
```
Comment (alexsnaps) [inline limitador/src/storage/mod.rs:124]: Is there a reason you made this a `Vec`? Cause it eventually gets sorted?

**Review by alexsnaps — COMMENTED** [review]
I had to tweak the bench to actually hit the new code path (loading the counters). I need to rerun it all on my linux machine tho, MacOS is too flaky on my runs, BUT I do seem to barely see any perf degradation... which makes me wonder if we should burn the other path and get rid of that `bool` altogether... wdyt?
Also, do we need to make this configurable? Is there a case where people would NOT want those headers to be present?

File: limitador/src/storage/in_memory.rs (hunk 1 above)
Comment (chirino) [reply limitador/src/storage/in_memory.rs:46]: what is the 'it' we should get rid of?

File: limitador/src/storage/mod.rs (hunk 2 above)
Comment (chirino) [reply limitador/src/storage/mod.rs:124]: and HashSet does not have a iter_mut() method.

File: limitador/src/storage/redis/redis_sync.rs [hunk 3]
```diff
@@ -157,6 +147,50 @@ impl RedisStorage {
     }
 }
 
+pub fn is_limited(
```
[the remark this answers is not in the document]
Comment (chirino) [reply limitador/src/storage/redis/redis_sync.rs:150]: will do.

File: limitador/src/storage/in_memory.rs [hunk 4]
```diff
@@ -31,22 +31,40 @@ impl CounterStorage for InMemoryStorage {
...
+                    return Ok(Authorization::Limited(
+                        counter.limit().name().map(|n| n.to_owned()),
+                    ));
+                }
             }
         }
 
-        for counter in counters {
-            self.insert_or_update_counter(&mut stored_counters, &counter, delta)
+        for counter in counters.iter() {
```
[the remark this answers is not in the document]
Comment (chirino) [reply limitador/src/storage/in_memory.rs:66]: yep.. will fix.

File: limitador-server/src/envoy_rls/server.rs [hunk 5]
```diff
@@ -110,6 +122,54 @@ impl RateLimitService for MyRateLimiter {
...
+            Ordering::Greater
+        }
+    });
+
+    let mut all_limits_text = String::with_capacity(20 * counters.len());
+    counters.iter_mut().for_each(|counter| {
+        all_limits_text
+            .push_str(format!(", {};w={}", counter.max_value(), counter.seconds()).as_str());
+        if let Some(name) = counter.limit().name() {
+            all_limits_text.push_str(format!(";name=\"{}\"", name).as_str());
```
[the remark this answers is not in the document]
Comment (chirino) [reply limitador-server/src/envoy_rls/server.rs:143]: Likely.. the spec BNF does not clarify this very well.  What are the escaping rules?  Maybe I should just translate all `"` chars to something else like `'`

File: limitador/src/storage/in_memory.rs (hunk 1 above)
Comment (alexsnaps) [reply limitador/src/storage/in_memory.rs:46]: `TtlCache`, but that's not for this PR...

File: limitador/src/storage/mod.rs (hunk 2 above)
Comment (alexsnaps) [reply limitador/src/storage/mod.rs:124]: Ah, yeah of course it doesn't... But given the performance improvement avoiding to dedup as we put them in a Set, I'd say this is fine, users "just" need to make sure they don't provide dups...

Comment (chirino) [reply limitador/src/storage/mod.rs:124]: Should we dedup in the counters_that_apply function?

Comment (alexsnaps) [reply limitador/src/storage/mod.rs:124]: No, I think the perf gain is worth it, it'd be weird someone feeds in dups anyways...

**Comment by alexsnaps** [comment]
What we now need is a way to configure the server to call into the limiter's `check_rate_limited_and_update_getting_counters` or `check_rate_limited_and_update` and set the headers conditionally as well. Based on my testing, in cases where you have multiple counters per request, this leads to a up to 20% latency degradation.

**Pull request merged**